e2e/main_test.go

```diff
@@ -16,53 +16,101 @@ package e2e
 
 import (
 	"fmt"
+	"io"
+	"net/http"
+	"net/http/httptest"
 	"os"
 	"path/filepath"
 	"runtime"
+	"strings"
 	"testing"
 )
 
 //nolint:golint
-const E2E_GETENVOY_BINARY = "E2E_GETENVOY_BINARY"
+const (
+	getenvoyBinaryEnvKey   = "E2E_GETENVOY_BINARY"
+	envoyVersionsUrlEnvKey = "ENVOY_VERSIONS_URL"
+	envoyVersionsJson      = "../site/envoy_versions.json"
+)
 
 // TestMain ensures the "getenvoy" binary is valid.
 func TestMain(m *testing.M) {
 	// As this is an e2e test, we execute all tests with a binary compiled earlier.
 	path, err := readGetEnvoyPath()
 	if err != nil {
-		fmt.Fprintf(os.Stderr, `failed to start e2e tests due to an invalid "getenvoy" binary: %v`, err)
-		os.Exit(1)
+		exitOnInvalidBinary(err)
 	}
 	getEnvoyPath = path
 
+	versionLine, _, err := getEnvoy("--version").exec()
+	if err != nil {
+		exitOnInvalidBinary(err)
+	}
+
+	if _, ok := os.LookupEnv(envoyVersionsUrlEnvKey); !ok && strings.Contains(versionLine, "SNAPSHOT") {
+		s, err := mockEnvoyVersionsServer()
```

Comment:
nit you forgot to close this mock server 😄

Reply:
linter yelled at me because we call os.Exit :)

Reply:
added a comment